### DISCUSSION TITLE
Interface inheritence and XAML compiler NRE

### DISCUSSION THREAD

**Discussion by IanRawley** [AvaloniaUI/Avalonia#14178] [Q&A]
I have a habit of making sets of Read Only and Read / Write interfaces (and an implementation) like the following:

```
public interface IReadableExample
{
    public string Name { get; }
}

public interface IWriteableExample : IReadableExample
{
    public new string Name { set; }
}

public class ExampleImpl : ReactiveObject, IWriteableExample
{
    private string _name = "Binding with Interfaces!";
    public string Name { get => _name; set => this.RaiseAndSetIfChanged(ref _name, value); }
}
```

This works fine in all situations I've encountered but 1, compiled binding to the getter via IWriteableExample, as in the following: 
```
public class MainViewModel : ViewModelBase
{
    public string Greeting => "Welcome to Avalonia!";

    private ExampleImpl _example = new();

    public IReadableExample ReadableExample { get => _example; }

    public IWriteableExample WriteableExample { get => _example; }
}

with MainView.xaml

<TextBox Text="{Binding WriteableExample.Name}" HorizontalAlignment="Center" VerticalAlignment="Center"/>


```
  In this case, trying to build in Avalonia has the XAML compiler throw the following:

```

Error	AVLN:0004	Internal compiler error while emitting node Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.XamlIlBindingPathHelper+XamlIlBindingPathNode:
System.NullReferenceException: Object reference not set to an instance of an object.
   at Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.XamlIlClrPropertyInfoEmitter.GetKey(IXamlProperty property, String indexerArgumentsKey) in /_/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlClrPropertyInfoHelper.cs:line 27
   at Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.XamlIlClrPropertyInfoEmitter.<>c__DisplayClass4_0.<Emit>g__Get|0() in /_/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlClrPropertyInfoHelper.cs:line 43
   at Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.XamlIlClrPropertyInfoEmitter.Emit(XamlEmitContext`2 context, IXamlILEmitter codeGen, IXamlProperty property, IEnumerable`1 indexerArguments, String indexerArgumentsKey) in /_/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlClrPropertyInfoHelper.cs:line 157
 
```

Now I understand this is probably because the IWriteableExample interface in this example doesn't explicitly have a Getter due to how C# handles interfaces, inheritance (and method, etc hiding), and properties, which the XAML Compiler doesn't like.  

It can be worked around easily by explicitly adding a Getter to the IWriteableExample, although I don't like this idea as then we are actually hiding the IReadableExample Getter, which could then be implemented explicitly and differently.
Another alternative workaround is to use ReflectionBinding rather than compiled, which works but then introduces (more) reflection which could cause trouble with trimming, etc in the future.

So my question here is: Is this a bug, and if so where would be best to report it?  The Avalonia repository, or the XAML compiler repository?  

[Here's a link to a minimal reproduction.](https://github.com/IanRawley/Interface_XAML_NRE_Example.git)

**Comment by timunie** [comment]
I guess `public new string Name { set; }` does hide the getter. Should be `{get; set; }` imo.

**Comment by maxkatz6** [comment]
Yes, it is by design, as there is no getter. The same code wouldn't work in pure C# as well (ignoring the fact that you can cast to IReadableExample simultaneously there).
What is a bug is the error message. It should be clear message about missing getter. 

This line blindly tries to access Getter:
https://github.com/AvaloniaUI/Avalonia/blob/8ea685eae40a9dd581bfd7ecf19f2f5604d21c17/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlClrPropertyInfoHelper.cs#L27

**Comment by IanRawley** [comment]
> Yes, it is by design, as there is no getter. The same code wouldn't work in pure C# as well (ignoring the fact that you can cast to IReadableExample simultaneously there). 

Basically what I figured, due to the requirement to hide the IReadableExample property with the new keyword.  I'm interested in what you mean by "The same code wouldn't work in pure C#", do you mean by skipping the XAML and creating the binding in C#?  

As I said, this works in all situations I've encountered except compiled bindings.  General usage of this pattern hasn't caused me any problems, and reflection bindings also work.  Is this sort of split Setter / Getter property inheritance something that should be able to be handled?
